### DISCUSSION TITLE
How to load/import/include FlexSearch into common environments

### DISCUSSION THREAD

**Discussion by ts-thomas** [nextapps-de/flexsearch#417] [Announcements]
Please let us collect all the environments which needs to be supported out of the box and what is needed to reach this goal. It should have the most simple solution for the specific environment. So please feel free to make your suggestions and don't forget to provide a **full** example of what is needed to do.

### Legacy ES5 Script Tags

```html
<script src="node_modules/flexsearch/dist/flexsearch.bundle.min.js"></script>
<script>

    // FlexSearch is available on window.FlexSearch
    // Access FlexSearch static methods via bundled export (static class methods of FlexSearch)

    const index = FlexSearch.Index(options);
    const document = FlexSearch.Document(options);
    const worker = FlexSearch.Worker(options);

</script>
```

### ES6 Modules (ESM)

```html
<script type="module">

    // FlexSearch is NOT available on window.FlexSearch
    // Access FlexSearch static methods by importing them explicitely

    import Index from "./node_modules/flexsearch/dist/module/index";
    import Document from "./node_modules/flexsearch/dist/module/document";
    import Worker from "./node_modules/flexsearch/dist/module/worker";

    const index = new Index(options);
    const document = new Document(options);
    const worker = new WorkerIndex(options);

</script>
```

### ES6 Bundled Module (ESM)

```html
<script type="module">

    // FlexSearch is NOT available on window.FlexSearch
    // Access FlexSearch static methods via bundled export (static class methods of FlexSearch)

    import FlexSearch from "./node_modules/flexsearch/dist/flexsearch.bundle.module.min.js";

    const index = FlexSearch.Index(options);
    const document = FlexSearch.Document(options);
    const worker = FlexSearch.Worker(options);

</script>
```

### Default Require

```js
const FlexSearch = require("flexsearch");

const index = new FlexSearch.Index(options);
const document = new FlexSearch.Document(options);
const worker = new FlexSearch.Worker(options);
```

Or:

```js
const { Index, Document, Worker } = require("flexsearch");

const index = new Index(options);
const document = new Document(options);
const worker = new Worker(options);
```

**Comment by thexeos** [comment]
Using Vite and with tsconfig set to `moduleResolution = bundler`, I had success with:

```ts
// @ts-expect-error there are no types expected here
import { Index, Document, Worker } from 'flexsearch/dist/flexsearch.bundle'
```

It's not ideal, but it means `require()` can be avoided to appease the linter and bundler.

**Reply by ts-thomas** [reply]
Have you tried `from 'flexsearch/dist/flexsearch.bundle.module.min'`? It's included in the latest build. It would be nice if you was able to remove the `@ts-expect-error` flag.

**Reply by ts-thomas** [reply]
You can give `from 'flexsearch'` a chance also

**Reply by thexeos** [reply]
You are absolutely right. With latest release it is possible to do:

```js
import FlexSearch from 'flexsearch'
```

However doing

```js
import { Document } from 'flexsearch'
```

will result in error:

![image](https://github.com/nextapps-de/flexsearch/assets/2260539/be2e8a41-dfaf-4148-811f-77098932ec67)

due to how the generated file only provides default export and none of the named ones:

![image](https://github.com/nextapps-de/flexsearch/assets/2260539/4b02d8cf-bf8e-4b5f-a26f-12a02ff5d7a7)

So the latest version of my code now uses:

```js
import FlexSearch from 'flexsearch'

new FlexSearch.Document()
```

**Reply by masylum** [reply]
this does not work with IndexedDb
